File: day3/part2/solution.clj
(ns solution
  (:require [clojure.string :as str]))

(defn find-max-subsequence [line k]
  (if (< (count line) k)
    (str/join (repeat k "0"))
    (loop [i 0
           current-start-index 0
           result-strings []]
      (if (= i k)
        (str/join result-strings)
        (let [remaining-to-find (- k i)
              search-end-index (- (count line) remaining-to-find)
              search-str (if (<= current-start-index search-end-index)
                           (subs line current-start-index (inc search-end-index))
                           "")]
          (if (empty? search-str)
            (str/join (concat result-strings (repeat remaining-to-find "0")))
            (let [best-digit-str (last (sort (map str search-str)))
                  best-digit-offset (.indexOf search-str best-digit-str)
                  best-digit-abs-index (+ current-start-index best-digit-offset)]
              (recur (inc i)
                     (inc best-digit-abs-index)
                     (conj result-strings best-digit-str)))))))))

(defn -main [& args]
  (let [input-path (str (.getParent (java.io.File. *file*)) "/input.txt")
        k 12
        lines (str/split-lines (slurp input-path))
        total-joltage (->> lines
                           (map str/trim)
                           (filter not-empty)
                           (map #(find-max-subsequence % k))
                           (map bigint)
                           (reduce +))]
    (println (str "The total output joltage is: " total-joltage))))

(-main)

File: day3/part2/solution.exs
defmodule Day3Part2 do
  def solve(input) do
    input
    |> String.split("\n", trim: true)
    |> Enum.map(&solve_line/1)
    |> Enum.sum()
  end

  defp solve_line(line) do
    if String.length(line) < 12 do
      0
    else
      n = String.length(line)
      k = 12
      to_remove = n - k
      
      {final_stack, _} = 
        line
        |> String.graphemes()
        |> Enum.reduce({[], to_remove}, fn digit, {stack, removals} ->
            {new_stack, new_removals} = pop_smaller(digit, stack, removals)
            {[digit | new_stack], new_removals}
        end)

      final_stack
      |> Enum.reverse()
      |> Enum.take(k)
      |> Enum.join()
      |> String.to_integer()
    end
  end

  defp pop_smaller(digit, [h|t], removals) when removals > 0 and digit > h do
    pop_smaller(digit, t, removals - 1)
  end
  defp pop_smaller(_digit, stack, removals) do
    {stack, removals}
  end
end

input = File.read!("day3/part2/input.txt")
result = Day3Part2.solve(input)
IO.puts("The new total output joltage is: #{result}")